service/history/workflowTaskHandlerCallbacks.go

```diff
@@ -564,7 +557,16 @@ Update_History_Loop:
 
 		if workflowTaskHeartbeatTimeout {
 			// at this point, update is successful, but we still return an error to client so that the worker will give up this workflow
-			return nil, serviceerror.NewNotFound(fmt.Sprintf("workflow task heartbeat timeout"))
+			return nil, serviceerror.NewNotFound("workflow task heartbeat timeout")
+		}
+
+		if workflowTaskFailedErr != nil {
+			switch workflowTaskFailedErr.causeErr.(type) {
```

Comment:
I don't like that we have logic about what error to return back to the caller spread across multiple files.  I think we should have the workflow task error logic only within workflowTaskHandler.go.